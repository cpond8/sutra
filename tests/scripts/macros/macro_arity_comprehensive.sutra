;; Comprehensive Macro Arity Testing
;; Tests systematic macro parameter handling to prevent variadic macro bugs
;; Covers all macro parameter patterns: fixed, variadic, and mixed

(print "=== MACRO ARITY TESTING ===")

;; Test 1: Fixed arity macros
(define (fixed_two a b) (list a b))
(print "Fixed arity macro (2 params):")
(print "Success case:")
(print (fixed_two 1 2))

(print "Expected failures:")
;; These will be tested in separate error files since they terminate execution

;; Test 2: Variadic macros
(define (variadic_any ...args) args)
(print "Variadic macro (any number of args):")
(print "Success cases:")
(print (variadic_any))           ; 0 args - valid
(print (variadic_any 1))         ; 1 arg - valid
(print (variadic_any 1 2 3))     ; many args - valid

;; Test 3: Mixed parameter macros
(define (mixed_params required ...optional) (list required optional))
(print "Mixed parameter macro (1 required + variadic):")
(print "Success cases:")
(print (mixed_params "req"))             ; minimum args
(print (mixed_params "req" "opt1"))      ; with optional
(print (mixed_params "req" "opt1" "opt2" "opt3")) ; many optional

;; Test 4: Complex nested macro calls
(define (outer ...args) (inner ...args))
(define (inner a b) (+ a b))
(print "Nested macro with parameter forwarding:")
(print "Success case:")
(print (outer 3 4))

;; Test 5: Spread operator usage
(define (spread_test ...items) (list ...items))
(print "Spread operator in macro body:")
(print "Success cases:")
(print (spread_test 1 2 3))

;; Test 6: Parameter binding edge cases
(define (param_binding x) x)
(print "Parameter binding:")
(print "Success cases:")
(print (param_binding "test"))
(print (param_binding 42))
(print (param_binding (list 1 2 3)))

(print "=== MACRO ARITY TESTING COMPLETE ===")
(print "All macro arity patterns tested successfully")