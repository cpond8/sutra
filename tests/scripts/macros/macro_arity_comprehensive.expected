=== MACRO ARITY TESTING ===
Fixed arity macro (2 params):
Success case:
(1 2)
Expected failures:
Variadic macro (any number of args):
Success cases:
()
Evaluation error: SutraError { kind: Eval(EvalError { message: "Arity mismatch in call to 'eval'\n\nExpected first element must be a symbol naming an atom arguments, but received 1.\n\nArguments provided (1):   1: number 1", expanded_code: "(eval 1)", original_code: None, suggestion: Some("Check the function documentation for correct usage") }), span: Some(Span { start: 691, end: 692 }) }